clash_lib/src/proxy/utils/test_utils/mod.rs

```diff
@@ -0,0 +1,158 @@
+    session::{Session, SocksAddr},
+};
+use tokio::{
+    io::{split, AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt},
+    join,
+    net::TcpListener,
+};
+use tracing::{debug, error};
+
+pub mod config_helper;
+pub mod consts;
+pub mod docker_runner;
+
+// TODO: add the thoroughput metrics
+pub async fn ping_pong_test(handler: Arc<dyn OutboundHandler>, port: u16) -> anyhow::Result<()> {
+    // PATH: our proxy handler -> proxy-server(container) -> target local server(127.0.0.1:port)
+
+    let sess = Session {
+        destination: ("127.0.0.1".to_owned(), port)
+            .try_into()
+            .unwrap_or_else(|_| panic!("")),
+        ..Default::default()
+    };
+
+    let (_, resolver) = config_helper::load_config().await?;
+
+    let listener = TcpListener::bind(format!("0.0.0.0:{}", port).as_str()).await?;
+
+    async fn destination_fn<T>(incoming: T) -> anyhow::Result<()>
+    where
+        T: AsyncRead + AsyncWrite,
+    {
+        // Use inbound_stream here
+        let (mut read_half, mut write_half) = split(incoming);
+        let chunk = "world";
+        let mut buf = vec![0; 5];
+
+        for _ in 0..100 {
+            read_half.read_exact(&mut buf).await?;
+            assert_eq!(&buf, b"hello");
+        }
+
+        for _ in 0..100 {
+            write_half.write_all(chunk.as_bytes()).await?;
+            write_half.flush().await?;
+        }
+        Ok(())
+    }
+
+    let target_local_server_handler = tokio::spawn(async move {
+        match listener.accept().await {
+            Ok((stream, _)) => match destination_fn(stream).await {
+                Ok(_) => {}
+                Err(e) => error!("Failed to serve: {}", e),
+            },
+            Err(e) => {
+                // Handle error e, log it, or ignore it
+                error!("Failed to accept connection: {}", e);
+            }
+        }
+        debug!("server task finished");
+    });
+
+    async fn proxy_fn(stream: Box<dyn ChainedStream>) -> anyhow::Result<()> {
+        let (mut read_half, mut write_half) = split(stream);
+
+        let chunk = "hello";
+        let mut buf = vec![0; 5];
+
+        for _ in 0..100 {
+            write_half.write_all(chunk.as_bytes()).await?;
+        }
+        write_half.flush().await?;
+        drop(write_half);
+
+        for _ in 0..100 {
+            read_half.read_exact(&mut buf).await?;
+            assert_eq!(buf, "world".as_bytes().to_owned());
+        }
+        drop(read_half);
+        Ok(())
+    }
+
+    let proxy_task = tokio::spawn(async move {
+        match handler.connect_stream(&sess, resolver).await {
+            Ok(stream) => match proxy_fn(stream).await {
+                Ok(_) => {}
+                Err(e) => error!("Failed to to proxy: {}", e),
+            },
+            Err(e) => error!("Failed to accept connection: {}", e),
+        }
+        debug!("proxy task finished");
+    });
+
+    let _ = join!(proxy_task, target_local_server_handler);
+
+    Ok(())
```

Comment:
would this ping_pong_test always return success result? is there a way for us to know when it's failed?